src/allocation_engine/interval_tree.rs

```diff
@@ -355,10 +357,101 @@ impl InnerNode {
 
     /// Updates the sub-tree to keep balance.
     #[allow(dead_code)]
-    fn update_node(mut self: Box<Self>) -> Box<Self> {
+    fn update_height_and_balance(mut self: Box<Self>) -> Box<Self> {
         self.update_cached_height();
         self.rotate()
     }
+
+    /// Insert a new node in the subtree. After the node is inserted the
+    /// tree will be balanced.
+    #[allow(dead_code)]
+    pub(crate) fn insert(
+        mut self: Box<Self>,
+        key: Range,
+        node_state: NodeState,
+    ) -> Result<Box<Self>> {
+        if self.key.overlaps(&key) && self.key != key {
+            return Err(Error::Overlap(key, self.key));
+        }
+        match self.key.cmp(&key) {
+            Ordering::Equal => {
```

Comment:
Is this  `Ordering::Equal` having the repeated judgment as to the above `self.key != key`?